// bim-fragment/fragment.js
import { BufferGeometry, Matrix4 } from 'three';
import { FragmentMesh } from './fragment-mesh';
import { Blocks } from './blocks';
import { BVH } from './bvh';
/*
 * Fragments can contain one or multiple Instances of one or multiple Blocks
 * Each Instance is identified by an instanceID (property of THREE.InstancedMesh)
 * Each Block identified by a blockID (custom bufferAttribute per vertex)
 * Both instanceId and blockId are unsigned integers starting at 0 and going up sequentially
 * A specific Block of a specific Instance is an Item, identified by an itemID
 *
 * For example:
 * Imagine a fragment mesh with 8 instances and 2 elements (16 items, identified from A to P)
 * It will have instanceIds from 0 to 8, and blockIds from 0 to 2
 * If we raycast it, we will get an instanceId and the index of the found triangle
 * We can use the index to get the blockId for that triangle
 * Combining instanceId and blockId using the elementMap will give us the itemId
 * The items will look like this:
 *
 *    [ A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P ]
 *
 *  Where the criteria to sort the items is the following (Y-axis is instance, X-axis is block):
 *
 *        A  C  E  G  I  K  M  O
 *        B  D  F  H  J  L  N  P
 * */
export class Fragment {
    constructor(geometry, material, count) {
        this.fragments = {};
        this.items = [];
        this.hiddenInstances = {};
        this.mesh = new FragmentMesh(geometry, material, count);
        this.id = this.mesh.uuid;
        this.capacity = count;
        this.blocks = new Blocks(this);
        BVH.apply(geometry);
    }
    dispose(disposeResources = true) {
        this.items = null;
        if (disposeResources) {
            this.mesh.material.forEach((mat) => mat.dispose());
            BVH.dispose(this.mesh.geometry);
            this.mesh.geometry.dispose();
        }
        this.mesh.dispose();
        this.mesh = null;
        this.disposeNestedFragments();
    }
    getItemID(instanceID, blockID) {
        const index = this.getItemIndex(instanceID, blockID);
        return this.items[index];
    }
    getInstanceAndBlockID(itemID) {
        const index = this.items.indexOf(itemID);
        const instanceID = this.getInstanceIDFromIndex(index);
        const blockID = index % this.blocks.count;
        return { instanceID, blockID };
    }
    getVertexBlockID(geometry, index) {
        return geometry.attributes.blockID.array[index];
    }
    getItemData(itemID) {
        const index = this.items.indexOf(itemID);
        const instanceID = Math.ceil(index / this.blocks.count);
        const blockID = index % this.blocks.count;
        return { instanceID, blockID };
    }
    getInstance(instanceID, matrix) {
        return this.mesh.getMatrixAt(instanceID, matrix);
    }
    setInstance(instanceID, items) {
        this.checkIfInstanceExist(instanceID);
        this.mesh.setMatrixAt(instanceID, items.transform);
        this.mesh.instanceMatrix.needsUpdate = true;
        if (items.ids) {
            this.saveItemsInMap(items.ids, instanceID);
        }
    }
    addInstances(items) {
        this.resizeCapacityIfNeeded(items.length);
        const start = this.mesh.count;
        this.mesh.count += items.length;
        for (let i = 0; i < items.length; i++) {
            this.setInstance(start + i, items[i]);
        }
    }
    removeInstances(itemsIDs) {
        if (this.mesh.count <= 1) {
            this.clear();
            return;
        }
        this.deleteAndRearrangeInstances(itemsIDs);
        this.mesh.count -= itemsIDs.length;
        this.mesh.instanceMatrix.needsUpdate = true;
    }
    clear() {
        this.mesh.clear();
        this.mesh.count = 0;
        this.items = [];
    }
    addFragment(id, material = this.mesh.material) {
        const newGeometry = this.initializeGeometry();
        if (material === this.mesh.material) {
            this.copyGroups(newGeometry);
        }
        const newFragment = new Fragment(newGeometry, material, this.capacity);
        newFragment.mesh.applyMatrix4(this.mesh.matrix);
        newFragment.mesh.updateMatrix();
        this.fragments[id] = newFragment;
        return this.fragments[id];
    }
    removeFragment(id) {
        const fragment = this.fragments[id];
        if (fragment) {
            fragment.dispose(false);
            delete this.fragments[id];
        }
    }
    resetVisibility() {
        if (this.blocks.count > 1) {
            this.blocks.reset();
        }
        else {
            const hiddenInstances = Object.keys(this.hiddenInstances);
            this.makeInstancesVisible(hiddenInstances);
            this.hiddenInstances = {};
        }
    }
    setVisibility(itemIDs, visible) {
        if (this.blocks.count > 1) {
            this.toggleBlockVisibility(visible, itemIDs);
            this.mesh.geometry.disposeBoundsTree();
            BVH.apply(this.mesh.geometry);
        }
        else {
            this.toggleInstanceVisibility(visible, itemIDs);
        }
    }
    resize(size) {
        var _a;
        const newMesh = this.createFragmentMeshWithNewSize(size);
        this.capacity = size;
        const oldMesh = this.mesh;
        (_a = oldMesh.parent) === null || _a === void 0 ? void 0 : _a.add(newMesh);
        oldMesh.removeFromParent();
        this.mesh = newMesh;
        oldMesh.dispose();
    }
    async export() {
        const geometryBuffer = await this.mesh.export();
        const geometry = new File([new Blob([geometryBuffer])], `${this.id}.glb`);
        const fragmentData = {
            matrices: Array.from(this.mesh.instanceMatrix.array),
            ids: this.items
        };
        const dataString = JSON.stringify(fragmentData);
        const data = new File([new Blob([dataString])], `${this.id}.json`);
        return { geometry, data };
    }
    copyGroups(newGeometry) {
        newGeometry.groups = [];
        for (const group of this.mesh.geometry.groups) {
            newGeometry.groups.push({ ...group });
        }
    }
    initializeGeometry() {
        const newGeometry = new BufferGeometry();
        newGeometry.setAttribute('position', this.mesh.geometry.attributes.position);
        newGeometry.setAttribute('normal', this.mesh.geometry.attributes.normal);
        newGeometry.setAttribute('blockID', this.mesh.geometry.attributes.blockID);
        newGeometry.setIndex(Array.from(this.mesh.geometry.index.array));
        return newGeometry;
    }
    saveItemsInMap(ids, instanceId) {
        this.checkBlockNumberValid(ids);
        let counter = 0;
        for (const id of ids) {
            const index = this.getItemIndex(instanceId, counter);
            this.items[index] = id;
            counter++;
        }
    }
    resizeCapacityIfNeeded(newSize) {
        const necessaryCapacity = newSize + this.mesh.count;
        if (necessaryCapacity > this.capacity) {
            this.resize(necessaryCapacity);
        }
    }
    createFragmentMeshWithNewSize(capacity) {
        const newMesh = new FragmentMesh(this.mesh.geometry, this.mesh.material, capacity);
        newMesh.count = this.mesh.count;
        return newMesh;
    }
    disposeNestedFragments() {
        const fragments = Object.values(this.fragments);
        for (let i = 0; i < fragments.length; i++) {
            fragments[i].dispose();
        }
        this.fragments = {};
    }
    checkBlockNumberValid(ids) {
        if (ids.length > this.blocks.count) {
            throw new Error(`You passed more items (${ids.length}) than blocks in this instance (${this.blocks.count})`);
        }
    }
    checkIfInstanceExist(index) {
        if (index > this.mesh.count) {
            throw new Error(`The given index (${index}) exceeds the instances in this fragment (${this.mesh.count})`);
        }
    }
    // Assigns the index of the removed instance to the last instance
    // F.e. let there be 6 instances: (A) (B) (C) (D) (E) (F)
    // If instance (C) is removed: -> (A) (B) (F) (D) (E)
    deleteAndRearrangeInstances(ids) {
        const deletedItems = [];
        for (const id of ids) {
            const deleted = this.deleteAndRearrange(id);
            if (deleted) {
                deletedItems.push(deleted);
            }
        }
        for (const id of ids) {
            delete this.hiddenInstances[id];
        }
        return deletedItems;
    }
    deleteAndRearrange(id) {
        const index = this.items.indexOf(id);
        if (index === -1)
            return null;
        this.mesh.count--;
        const isLastElement = index === this.mesh.count;
        const instanceId = this.getInstanceIDFromIndex(index);
        const tempMatrix = new Matrix4();
        const transform = new Matrix4();
        this.mesh.getMatrixAt(instanceId, transform);
        if (isLastElement) {
            this.items.pop();
            return { ids: [id], transform };
        }
        const lastElement = this.mesh.count;
        this.items[index] = this.items[lastElement];
        this.items.pop();
        this.mesh.getMatrixAt(lastElement, tempMatrix);
        this.mesh.setMatrixAt(instanceId, tempMatrix);
        this.mesh.instanceMatrix.needsUpdate = true;
        return { ids: [id], transform };
    }
    getItemIndex(instanceId, blockId) {
        return instanceId * this.blocks.count + blockId;
    }
    getInstanceIDFromIndex(itemIndex) {
        return Math.trunc(itemIndex / this.blocks.count);
    }
    toggleInstanceVisibility(visible, itemIDs) {
        if (visible) {
            this.makeInstancesVisible(itemIDs);
        }
        else {
            this.makeInstancesInvisible(itemIDs);
        }
    }
    makeInstancesInvisible(itemIDs) {
        itemIDs = this.filterHiddenItems(itemIDs, false);
        const deletedItems = this.deleteAndRearrangeInstances(itemIDs);
        for (const item of deletedItems) {
            if (item.ids) {
                this.hiddenInstances[item.ids[0]] = item;
            }
        }
    }
    makeInstancesVisible(itemIDs) {
        const items = [];
        itemIDs = this.filterHiddenItems(itemIDs, true);
        for (const id of itemIDs) {
            items.push(this.hiddenInstances[id]);
            delete this.hiddenInstances[id];
        }
        this.addInstances(items);
    }
    filterHiddenItems(itemIDs, hidden) {
        const hiddenItems = Object.keys(this.hiddenInstances);
        return itemIDs.filter((item) => hidden ? hiddenItems.includes(item) : !hiddenItems.includes(item));
    }
    toggleBlockVisibility(visible, itemIDs) {
        const blockIDs = itemIDs.map((id) => this.getInstanceAndBlockID(id).blockID);
        if (visible) {
            this.blocks.add(blockIDs, false);
        }
        else {
            this.blocks.remove(blockIDs);
        }
    }
}
